// @sentry/serverless/cjs/gcpfunction/index.js
Object.defineProperty(exports, '__esModule', { value: true });

const node = require('@sentry/node');
const googleCloudGrpc = require('../google-cloud-grpc.js');
const googleCloudHttp = require('../google-cloud-http.js');
const http = require('./http.js');
const events = require('./events.js');
const cloud_events = require('./cloud_events.js');

/** @deprecated Use `getDefaultIntegrations(options)` instead. */
const defaultIntegrations = [
  // eslint-disable-next-line deprecation/deprecation
  ...node.defaultIntegrations,
  googleCloudHttp.googleCloudHttpIntegration({ optional: true }), // We mark this integration optional since '@google-cloud/common' module could be missing.
  googleCloudGrpc.googleCloudGrpcIntegration({ optional: true }), // We mark this integration optional since 'google-gax' module could be missing.
];

/** Get the default integrations for the GCP SDK. */
function getDefaultIntegrations(options) {
  return [
    ...node.getDefaultIntegrations(options),
    googleCloudHttp.googleCloudHttpIntegration({ optional: true }), // We mark this integration optional since '@google-cloud/common' module could be missing.
    googleCloudGrpc.googleCloudGrpcIntegration({ optional: true }), // We mark this integration optional since 'google-gax' module could be missing.
  ];
}

/**
 * @see {@link Sentry.init}
 */
function init(options = {}) {
  const opts = {
    _metadata: {} ,
    defaultIntegrations: getDefaultIntegrations(options),
    ...options,
  };

  opts._metadata.sdk = opts._metadata.sdk || {
    name: 'sentry.javascript.serverless',
    integrations: ['GCPFunction'],
    packages: [
      {
        name: 'npm:@sentry/serverless',
        version: node.SDK_VERSION,
      },
    ],
    version: node.SDK_VERSION,
  };

  node.init(opts);
}

exports.wrapHttpFunction = http.wrapHttpFunction;
exports.wrapEventFunction = events.wrapEventFunction;
exports.wrapCloudEventFunction = cloud_events.wrapCloudEventFunction;
exports.defaultIntegrations = defaultIntegrations;
exports.getDefaultIntegrations = getDefaultIntegrations;
exports.init = init;
