// @sentry/serverless/esm/awslambda.js
import { _optionalChain } from '@sentry/utils';
import { existsSync } from 'fs';
import { hostname } from 'os';
import { basename, resolve } from 'path';
import { types } from 'util';
import * as node from '@sentry/node';
import { defaultIntegrations as defaultIntegrations$1, getDefaultIntegrations as getDefaultIntegrations$1, SDK_VERSION, init as init$1, withScope, captureMessage, continueTrace, startSpanManual, getCurrentScope, captureException, flush } from '@sentry/node';
export * from '@sentry/node';
import { logger, isString } from '@sentry/utils';
import { performance } from 'perf_hooks';
import { SEMANTIC_ATTRIBUTE_SENTRY_SOURCE, SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN } from '@sentry/core';
import { awsServicesIntegration } from './awsservices.js';
import { DEBUG_BUILD } from './debug-build.js';
import { markEventUnhandled } from './utils.js';

const { isPromise } = types;

// https://www.npmjs.com/package/aws-lambda-consumer

/** @deprecated Use `getDefaultIntegrations(options)` instead. */
const defaultIntegrations = [
  // eslint-disable-next-line deprecation/deprecation
  ...defaultIntegrations$1,
  awsServicesIntegration({ optional: true }),
];

/** Get the default integrations for the AWSLambda SDK. */
function getDefaultIntegrations(options) {
  return [...getDefaultIntegrations$1(options), awsServicesIntegration({ optional: true })];
}

/**
 * Initializes the Sentry AWS Lambda SDK.
 *
 * @param options Configuration options for the SDK, @see {@link AWSLambdaOptions}.
 */
function init(options = {}) {
  const opts = {
    _metadata: {} ,
    defaultIntegrations: getDefaultIntegrations(options),
    ...options,
  };

  opts._metadata.sdk = opts._metadata.sdk || {
    name: 'sentry.javascript.serverless',
    integrations: ['AWSLambda'],
    packages: [
      {
        name: 'npm:@sentry/serverless',
        version: SDK_VERSION,
      },
    ],
    version: SDK_VERSION,
  };

  init$1(opts);
}

/** */
function tryRequire(taskRoot, subdir, mod) {
  const lambdaStylePath = resolve(taskRoot, subdir, mod);
  if (existsSync(lambdaStylePath) || existsSync(`${lambdaStylePath}.js`)) {
    // Lambda-style path
    return require(lambdaStylePath);
  }
  // Node-style path
  return require(require.resolve(mod, { paths: [taskRoot, subdir] }));
}

/** */
function isPromiseAllSettledResult(result) {
  return result.every(
    v =>
      Object.prototype.hasOwnProperty.call(v, 'status') &&
      (Object.prototype.hasOwnProperty.call(v, 'value') || Object.prototype.hasOwnProperty.call(v, 'reason')),
  );
}

/** */
function getRejectedReasons(results) {
  return results.reduce((rejected, result) => {
    if (result.status === 'rejected' && result.reason) rejected.push(result.reason);
    return rejected;
  }, []);
}

/** */
function tryPatchHandler(taskRoot, handlerPath) {

  const handlerDesc = basename(handlerPath);
  const match = handlerDesc.match(/^([^.]*)\.(.*)$/);
  if (!match) {
    DEBUG_BUILD && logger.error(`Bad handler ${handlerDesc}`);
    return;
  }

  const [, handlerMod, handlerName] = match;

  let obj;
  try {
    const handlerDir = handlerPath.substring(0, handlerPath.indexOf(handlerDesc));
    obj = tryRequire(taskRoot, handlerDir, handlerMod);
  } catch (e) {
    DEBUG_BUILD && logger.error(`Cannot require ${handlerPath} in ${taskRoot}`, e);
    return;
  }

  let mod;
  let functionName;
  handlerName.split('.').forEach(name => {
    mod = obj;
    obj = obj && (obj )[name];
    functionName = name;
  });
  if (!obj) {
    DEBUG_BUILD && logger.error(`${handlerPath} is undefined or not exported`);
    return;
  }
  if (typeof obj !== 'function') {
    DEBUG_BUILD && logger.error(`${handlerPath} is not a function`);
    return;
  }

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  (mod )[functionName] = wrapHandler(obj);
}

/**
 * Tries to invoke context.getRemainingTimeInMillis if not available returns 0
 * Some environments use AWS lambda but don't support this function
 * @param context
 */
function tryGetRemainingTimeInMillis(context) {
  return typeof context.getRemainingTimeInMillis === 'function' ? context.getRemainingTimeInMillis() : 0;
}

/**
 * Adds additional information from the environment and AWS Context to the Sentry Scope.
 *
 * @param scope Scope that should be enhanced
 * @param context AWS Lambda context that will be used to extract some part of the data
 * @param startTime performance.now() when wrapHandler was invoked
 */
function enhanceScopeWithEnvironmentData(scope, context, startTime) {
  scope.setContext('aws.lambda', {
    aws_request_id: context.awsRequestId,
    function_name: context.functionName,
    function_version: context.functionVersion,
    invoked_function_arn: context.invokedFunctionArn,
    execution_duration_in_millis: performance.now() - startTime,
    remaining_time_in_millis: tryGetRemainingTimeInMillis(context),
    'sys.argv': process.argv,
  });

  scope.setContext('aws.cloudwatch.logs', {
    log_group: context.logGroupName,
    log_stream: context.logStreamName,
    url: `https://console.aws.amazon.com/cloudwatch/home?region=${
      process.env.AWS_REGION
    }#logsV2:log-groups/log-group/${encodeURIComponent(context.logGroupName)}/log-events/${encodeURIComponent(
      context.logStreamName,
    )}?filterPattern="${context.awsRequestId}"`,
  });
}

/**
 * Adds additional transaction-related information from the environment and AWS Context to the Sentry Scope.
 *
 * @param scope Scope that should be enhanced
 * @param context AWS Lambda context that will be used to extract some part of the data
 */
function enhanceScopeWithTransactionData(scope, context) {
  scope.addEventProcessor(event => {
    event.transaction = context.functionName;
    return event;
  });
  scope.setTag('server_name', process.env._AWS_XRAY_DAEMON_ADDRESS || process.env.SENTRY_NAME || hostname());
  scope.setTag('url', `awslambda:///${context.functionName}`);
}

/**
 * Wraps a lambda handler adding it error capture and tracing capabilities.
 *
 * @param handler Handler
 * @param options Options
 * @returns Handler
 */
function wrapHandler(
  handler,
  wrapOptions = {},
) {
  const START_TIME = performance.now();
  const options = {
    flushTimeout: 2000,
    callbackWaitsForEmptyEventLoop: false,
    captureTimeoutWarning: true,
    timeoutWarningLimit: 500,
    captureAllSettledReasons: false,
    startTrace: true,
    ...wrapOptions,
  };
  let timeoutWarningTimer;

  // AWSLambda is like Express. It makes a distinction about handlers based on its last argument
  // async (event) => async handler
  // async (event, context) => async handler
  // (event, context, callback) => sync handler
  // Nevertheless whatever option is chosen by user, we convert it to async handler.
  const asyncHandler =
    handler.length > 2
      ? (event, context) =>
          new Promise((resolve, reject) => {
            const rv = (handler )(event, context, (error, result) => {
              if (error === null || error === undefined) {
                resolve(result); // eslint-disable-line @typescript-eslint/no-non-null-assertion
              } else {
                reject(error);
              }
            }) ;

            // This should never happen, but still can if someone writes a handler as
            // `async (event, context, callback) => {}`
            if (isPromise(rv)) {
              void (rv ).then(resolve, reject);
            }
          })
      : (handler );

  return async (event, context) => {
    context.callbackWaitsForEmptyEventLoop = options.callbackWaitsForEmptyEventLoop;

    // In seconds. You cannot go any more granular than this in AWS Lambda.
    const configuredTimeout = Math.ceil(tryGetRemainingTimeInMillis(context) / 1000);
    const configuredTimeoutMinutes = Math.floor(configuredTimeout / 60);
    const configuredTimeoutSeconds = configuredTimeout % 60;

    const humanReadableTimeout =
      configuredTimeoutMinutes > 0
        ? `${configuredTimeoutMinutes}m${configuredTimeoutSeconds}s`
        : `${configuredTimeoutSeconds}s`;

    // When `callbackWaitsForEmptyEventLoop` is set to false, which it should when using `captureTimeoutWarning`,
    // we don't have a guarantee that this message will be delivered. Because of that, we don't flush it.
    if (options.captureTimeoutWarning) {
      const timeoutWarningDelay = tryGetRemainingTimeInMillis(context) - options.timeoutWarningLimit;

      timeoutWarningTimer = setTimeout(() => {
        withScope(scope => {
          scope.setTag('timeout', humanReadableTimeout);
          captureMessage(`Possible function timeout: ${context.functionName}`, 'warning');
        });
      }, timeoutWarningDelay) ;
    }

    async function processResult(span) {
      const scope = getCurrentScope();

      let rv;
      try {
        enhanceScopeWithEnvironmentData(scope, context, START_TIME);

        rv = await asyncHandler(event, context);

        // We manage lambdas that use Promise.allSettled by capturing the errors of failed promises
        if (options.captureAllSettledReasons && Array.isArray(rv) && isPromiseAllSettledResult(rv)) {
          const reasons = getRejectedReasons(rv);
          reasons.forEach(exception => {
            captureException(exception, scope => markEventUnhandled(scope));
          });
        }
      } catch (e) {
        captureException(e, scope => markEventUnhandled(scope));
        throw e;
      } finally {
        clearTimeout(timeoutWarningTimer);
        _optionalChain([span, 'optionalAccess', _ => _.end, 'call', _2 => _2()]);
        await flush(options.flushTimeout).catch(e => {
          DEBUG_BUILD && logger.error(e);
        });
      }
      return rv;
    }

    if (options.startTrace) {
      const eventWithHeaders = event ;

      const sentryTrace =
        eventWithHeaders.headers && isString(eventWithHeaders.headers['sentry-trace'])
          ? eventWithHeaders.headers['sentry-trace']
          : undefined;
      const baggage = _optionalChain([eventWithHeaders, 'access', _3 => _3.headers, 'optionalAccess', _4 => _4.baggage]);

      return continueTrace({ sentryTrace, baggage }, () => {
        return startSpanManual(
          {
            name: context.functionName,
            op: 'function.aws.lambda',
            attributes: {
              [SEMANTIC_ATTRIBUTE_SENTRY_SOURCE]: 'component',
              [SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: 'auto.function.serverless',
            },
          },
          span => {
            enhanceScopeWithTransactionData(getCurrentScope(), context);

            return processResult(span);
          },
        );
      });
    }

    return withScope(async () => {
      return processResult(undefined);
    });
  };
}

export { defaultIntegrations, getDefaultIntegrations, init, tryPatchHandler, wrapHandler };
