// @sentry/serverless/cjs/index.js
Object.defineProperty(exports, '__esModule', { value: true });

const awslambda = require('./awslambda.js');
const index = require('./gcpfunction/index.js');
const awsservices = require('./awsservices.js');
const node = require('@sentry/node');



exports.AWSLambda = awslambda;
exports.GCPFunction = index;
exports.AWSServices = awsservices.AWSServices;
exports.awsServicesIntegration = awsservices.awsServicesIntegration;
exports.DEFAULT_USER_INCLUDES = node.DEFAULT_USER_INCLUDES;
exports.Handlers = node.Handlers;
exports.Hub = node.Hub;
exports.Integrations = node.Integrations;
exports.NodeClient = node.NodeClient;
exports.SDK_VERSION = node.SDK_VERSION;
exports.SEMANTIC_ATTRIBUTE_SENTRY_OP = node.SEMANTIC_ATTRIBUTE_SENTRY_OP;
exports.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN = node.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN;
exports.SEMANTIC_ATTRIBUTE_SENTRY_SAMPLE_RATE = node.SEMANTIC_ATTRIBUTE_SENTRY_SAMPLE_RATE;
exports.SEMANTIC_ATTRIBUTE_SENTRY_SOURCE = node.SEMANTIC_ATTRIBUTE_SENTRY_SOURCE;
exports.Scope = node.Scope;
exports.addBreadcrumb = node.addBreadcrumb;
exports.addEventProcessor = node.addEventProcessor;
exports.addGlobalEventProcessor = node.addGlobalEventProcessor;
exports.addIntegration = node.addIntegration;
exports.addRequestDataToEvent = node.addRequestDataToEvent;
exports.anrIntegration = node.anrIntegration;
exports.autoDiscoverNodePerformanceMonitoringIntegrations = node.autoDiscoverNodePerformanceMonitoringIntegrations;
exports.captureCheckIn = node.captureCheckIn;
exports.captureConsoleIntegration = node.captureConsoleIntegration;
exports.captureEvent = node.captureEvent;
exports.captureException = node.captureException;
exports.captureMessage = node.captureMessage;
exports.captureSession = node.captureSession;
exports.close = node.close;
exports.configureScope = node.configureScope;
exports.consoleIntegration = node.consoleIntegration;
exports.contextLinesIntegration = node.contextLinesIntegration;
exports.continueTrace = node.continueTrace;
exports.createGetModuleFromFilename = node.createGetModuleFromFilename;
exports.createTransport = node.createTransport;
exports.debugIntegration = node.debugIntegration;
exports.dedupeIntegration = node.dedupeIntegration;
exports.deepReadDirSync = node.deepReadDirSync;
exports.defaultIntegrations = node.defaultIntegrations;
exports.defaultStackParser = node.defaultStackParser;
exports.endSession = node.endSession;
exports.extraErrorDataIntegration = node.extraErrorDataIntegration;
exports.extractRequestData = node.extractRequestData;
exports.extractTraceparentData = node.extractTraceparentData;
exports.flush = node.flush;
exports.functionToStringIntegration = node.functionToStringIntegration;
exports.getActiveSpan = node.getActiveSpan;
exports.getActiveTransaction = node.getActiveTransaction;
exports.getClient = node.getClient;
exports.getCurrentHub = node.getCurrentHub;
exports.getCurrentScope = node.getCurrentScope;
exports.getDefaultIntegrations = node.getDefaultIntegrations;
exports.getGlobalScope = node.getGlobalScope;
exports.getHubFromCarrier = node.getHubFromCarrier;
exports.getIsolationScope = node.getIsolationScope;
exports.getModuleFromFilename = node.getModuleFromFilename;
exports.getSentryRelease = node.getSentryRelease;
exports.getSpanStatusFromHttpCode = node.getSpanStatusFromHttpCode;
exports.hapiIntegration = node.hapiIntegration;
exports.httpClientIntegration = node.httpClientIntegration;
exports.httpIntegration = node.httpIntegration;
exports.inboundFiltersIntegration = node.inboundFiltersIntegration;
exports.init = node.init;
exports.isInitialized = node.isInitialized;
exports.lastEventId = node.lastEventId;
exports.linkedErrorsIntegration = node.linkedErrorsIntegration;
exports.localVariablesIntegration = node.localVariablesIntegration;
exports.makeMain = node.makeMain;
exports.makeNodeTransport = node.makeNodeTransport;
exports.metrics = node.metrics;
exports.modulesIntegration = node.modulesIntegration;
exports.nativeNodeFetchintegration = node.nativeNodeFetchintegration;
exports.nodeContextIntegration = node.nodeContextIntegration;
exports.onUncaughtExceptionIntegration = node.onUncaughtExceptionIntegration;
exports.onUnhandledRejectionIntegration = node.onUnhandledRejectionIntegration;
exports.parameterize = node.parameterize;
exports.reportingObserverIntegration = node.reportingObserverIntegration;
exports.requestDataIntegration = node.requestDataIntegration;
exports.rewriteFramesIntegration = node.rewriteFramesIntegration;
exports.runWithAsyncContext = node.runWithAsyncContext;
exports.sessionTimingIntegration = node.sessionTimingIntegration;
exports.setContext = node.setContext;
exports.setCurrentClient = node.setCurrentClient;
exports.setExtra = node.setExtra;
exports.setExtras = node.setExtras;
exports.setHttpStatus = node.setHttpStatus;
exports.setMeasurement = node.setMeasurement;
exports.setTag = node.setTag;
exports.setTags = node.setTags;
exports.setUser = node.setUser;
exports.spanStatusfromHttpCode = node.spanStatusfromHttpCode;
exports.spotlightIntegration = node.spotlightIntegration;
exports.startActiveSpan = node.startActiveSpan;
exports.startInactiveSpan = node.startInactiveSpan;
exports.startSession = node.startSession;
exports.startSpan = node.startSpan;
exports.startSpanManual = node.startSpanManual;
exports.startTransaction = node.startTransaction;
exports.trpcMiddleware = node.trpcMiddleware;
exports.withActiveSpan = node.withActiveSpan;
exports.withIsolationScope = node.withIsolationScope;
exports.withMonitor = node.withMonitor;
exports.withScope = node.withScope;
